File: testutils-logrecorder-logback/src/main/java/info/novatec/testit/testutils/logrecorder/junit5/LogbackRecorderExtension.java
package info.novatec.testit.testutils.logrecorder.junit5;

import static java.util.Arrays.stream;
import static java.util.stream.Collectors.toList;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.jupiter.api.extension.TestExtensionContext;
import org.slf4j.LoggerFactory;

import ch.qos.logback.classic.Level;
import ch.qos.logback.classic.Logger;

import info.novatec.testit.testutils.logrecorder.api.LogRecord;
import info.novatec.testit.testutils.logrecorder.internal.AbstractLogRecorderExtension;
import info.novatec.testit.testutils.logrecorder.internal.logback.LogbackLog;
import info.novatec.testit.testutils.logrecorder.internal.logback.LogbackLogRecorder;


/**
 * This extension will record the loggers specified by a {@link RecordLoggers} annotation and inject the {@link LogRecord}
 * as
 * a parameter into the test method.
 * <p>
 * Recording a logger will set its log level to {@link Level} to {@link Level#ALL} for the duration of the test. After the
 * test was executed, the log level will be restored to whatever it was before.
 *
 * @see RecordLoggers
 * @see LogRecord
 */
public class LogbackRecorderExtension extends AbstractLogRecorderExtension {

    private static final String RECORDERS = "recorders";
    private static final String ANNOTATION_NOT_FOUND_MSG =
        "no @" + RecordLoggers.class.getSimpleName() + " annotation found on test method!";

    @Override
    public void beforeTestExecution(TestExtensionContext context) {
        initExtensionContext(context);
        getRecorders(context).forEach(LogbackLogRecorder::start);
    }

    @Override
    public void afterTestExecution(TestExtensionContext context) {
        getRecorders(context).forEach(LogbackLogRecorder::stop);
    }

    private void initExtensionContext(TestExtensionContext context) {
        LogbackLog log = new LogbackLog();
        storeLog(context, log);
        List<LogbackLogRecorder> recorders = getLoggers(context)//
            .map(logger -> new LogbackLogRecorder(logger, log))//
            .collect(toList());
        storeRecorders(context, recorders);
    }

    private Stream<Logger> getLoggers(ExtensionContext context) {
        RecordLoggers annotation = context.getTestMethod()
            .map(method -> method.getAnnotation(RecordLoggers.class))
            .orElseThrow(() -> new IllegalStateException(ANNOTATION_NOT_FOUND_MSG));
        List<org.slf4j.Logger> combinedLoggers = new ArrayList<>();
        stream(annotation.value()).map(LoggerFactory::getLogger).forEach(combinedLoggers::add);
        stream(annotation.names()).map(LoggerFactory::getLogger).forEach(combinedLoggers::add);
        return combinedLoggers.stream().map(logger -> ( Logger ) logger);
    }

    private void storeRecorders(ExtensionContext context, List<LogbackLogRecorder> loggerData) {
        getStore(context).put(RECORDERS, loggerData);
    }

    @SuppressWarnings("unchecked")
    private List<LogbackLogRecorder> getRecorders(ExtensionContext context) {
        return ( List<LogbackLogRecorder> ) getStore(context).get(RECORDERS);
    }

}

File: logrecorder/logrecorder-logback/src/main/kotlin/info/novatec/testit/logrecorder/logback/junit5/LogbackRecorderExtension.kt
/*
 * Copyright 2017-2021 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package info.novatec.testit.logrecorder.logback.junit5

import ch.qos.logback.classic.Level
import ch.qos.logback.classic.Logger
import info.novatec.testit.logrecorder.api.LogRecord
import info.novatec.testit.logrecorder.logback.LogbackLogRecord
import info.novatec.testit.logrecorder.logback.LogbackLogRecorder
import org.junit.jupiter.api.extension.*
import org.junit.jupiter.api.extension.ExtensionContext.Namespace
import org.junit.jupiter.api.extension.ExtensionContext.Store
import org.slf4j.LoggerFactory

/**
 * This extension will record the loggers specified by a [RecordLoggers] annotation and inject the [LogRecord]
 * as a parameter into the test method.
 *
 * Recording a logger will set its log level to to [Level.ALL] for the duration of the test. After the
 * test was executed, the log level will be restored to whatever it was before.
 *
 * @see RecordLoggers
 * @see LogRecord
 * @since 1.0
 */
class LogbackRecorderExtension : BeforeTestExecutionCallback, AfterTestExecutionCallback, ParameterResolver {

    private val namespace = Namespace.create(javaClass)

    override fun supportsParameter(parameterContext: ParameterContext, extensionContext: ExtensionContext): Boolean {
        return LogRecord::class.java == parameterContext.parameter.type
    }

    override fun resolveParameter(parameterContext: ParameterContext, extensionContext: ExtensionContext): Any {
        return extensionContext.logRecord
    }

    override fun beforeTestExecution(context: ExtensionContext) {
        val annotation = context.requiredTestMethod.getAnnotation(RecordLoggers::class.java)
            ?: error("no @RecordLoggers annotation found on test method!")
        val fromClasses = annotation.value.map { LoggerFactory.getLogger(it.java) }
        val fromNames = annotation.names.map { LoggerFactory.getLogger(it) }

        val logRecord = LogbackLogRecord()
        context.logRecord = logRecord

        val recorders = (fromClasses + fromNames)
            .distinct()
            .map { it as Logger }
            .map { LogbackLogRecorder(it, logRecord) }
            .onEach { it.start() }
        context.recorders = recorders
    }

    override fun afterTestExecution(context: ExtensionContext) {
        context.recorders.forEach { it.stop() }
    }

    private var ExtensionContext.recorders: List<LogbackLogRecorder>
        get() = store.get("log-recorders") as List<LogbackLogRecorder>
        set(value) = store.put("log-recorders", value)

    private var ExtensionContext.logRecord: LogbackLogRecord
        get() = store.get("log-record", LogbackLogRecord::class.java)
        set(value) = store.put("log-record", value)

    private val ExtensionContext.store: Store
        get() = getStore(namespace)

}
